File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroInputValue.java
package com.xs0.gqlktx.schema.intro;

import com.xs0.gqlktx.ann.GraphQLObject;
import com.xs0.gqlktx.types.gql.GArgument;
import com.xs0.gqlktx.types.gql.GField;

@GraphQLObject("__InputValue")
public class GqlIntroInputValue {
    private final String name;
    private final String description;
    private final GqlIntroType type;
    private final String defaultValue;

    public GqlIntroInputValue(String name, String description, GqlIntroType type, String defaultValue) {
        this.name = name;
        this.description = description;
        this.type = type;
        this.defaultValue = defaultValue;
    }

    public GqlIntroInputValue(GArgument arg) {
        this.name = arg.getName();
        this.description = null; // TODO
        this.type = arg.getType().introspector();
        this.defaultValue = arg.getDefaultValue() != null ? arg.getDefaultValue().toString() : null;
    }

    public GqlIntroInputValue(GField value) {
        this.name = value.getName();
        this.description = null; // TODO
        this.type = value.getType().introspector();
        this.defaultValue = null;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public GqlIntroType getType() {
        return type;
    }

    public String getDefaultValue() {
        return defaultValue;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/schema/intro/GqlIntroInputValue.kt
package com.xs0.gqlktx.schema.intro

import com.xs0.gqlktx.GqlObject
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.types.gql.GArgument
import com.xs0.gqlktx.types.gql.GField
import com.xs0.gqlktx.types.gql.GInputField

@GqlObject("__InputValue")
class GqlIntroInputValue {
    val name: String
    val description: String?
    val type: GqlIntroType
    private val _defaultValue: Value?
    val defaultValue: String?
        get() = _defaultValue?.toString()

    constructor(name: String, description: String, type: GqlIntroType, _defaultValue: Value?) {
        this.name = name
        this.description = description
        this.type = type
        this._defaultValue = _defaultValue
    }

    constructor(arg: GArgument) {
        this.name = arg.name
        this.description = arg.description
        this.type = arg.type.introspector
        this._defaultValue = arg.defaultValue
    }

    constructor(value: GInputField) {
        this.name = value.name
        this.description = value.description
        this.type = value.type.introspector
        this._defaultValue = value.defaultValue
    }
}
